// @passbase/node/lib/configuration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PassbaseConfiguration = exports.SDK_NAME = exports.ResponseFormats = void 0;
const utils_1 = require("./utils");
// @ts-ignore
const package_json_1 = require("../package.json");
var ResponseFormats;
(function (ResponseFormats) {
    ResponseFormats["Json"] = "application/json";
    ResponseFormats["Xml"] = "application/xml";
})(ResponseFormats = exports.ResponseFormats || (exports.ResponseFormats = {}));
exports.SDK_NAME = `node@${package_json_1.version}`;
class PassbaseConfiguration {
    constructor(param) {
        if (!param.apiKey) {
            throw new utils_1.PassbaseError("missing apiKey");
        }
        if (param.format &&
            !Object.values(ResponseFormats).includes(param.format)) {
            throw new utils_1.PassbaseError(`unsupported response format`);
        }
        this.apiKey = param.apiKey;
        this.format = param.format || ResponseFormats.Json;
        this.basePath =
            param.basePath || "https://api.passbase.com/verification/v1";
    }
}
exports.PassbaseConfiguration = PassbaseConfiguration;
