// @passbase/node/lib/models/Identity.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Identity = void 0;
const IdentityOwner_1 = require("./IdentityOwner");
const IdentityResource_1 = require("./IdentityResource");
const Watchlist_1 = require("./Watchlist");
var IdentityStatus;
(function (IdentityStatus) {
    IdentityStatus["Created"] = "created";
    IdentityStatus["Processing"] = "processing";
    IdentityStatus["Processed"] = "processed";
    IdentityStatus["Verified"] = "verified";
    IdentityStatus["Unverified"] = "unverified";
    IdentityStatus["Pending"] = "pending";
    IdentityStatus["Approved"] = "approved";
    IdentityStatus["Declined"] = "declined";
})(IdentityStatus || (IdentityStatus = {}));
class Identity {
    constructor(obj) {
        this.id = obj.id;
        this.status = obj.status;
        this.created = new Date(obj.created * 1000);
        this.updated = new Date(obj.updated * 1000);
        this.owner = obj.owner ? new IdentityOwner_1.IdentityOwner(obj.owner) : null;
        this.score = obj.score;
        this.resources = (obj.resources || []).map((resource) => new IdentityResource_1.IdentityResource(resource));
        this.watchlist = obj.watchlist ? new Watchlist_1.Watchlist(obj.watchlist) : null;
    }
}
exports.Identity = Identity;
